source/FreeRTOS_IPv4.c

```diff
@@ -314,7 +314,6 @@ enum eFrameProcessingResult prvAllowIPPacketIPv4( const struct xIP_PACKET * cons
         }
         else if(
             ( FreeRTOS_FindEndPointOnIP_IPv4( ulDestinationIPAddress, 4 ) == NULL ) &&
-            ( pxNetworkBuffer->pxEndPoint == NULL ) &&
             /* Is it an IPv4 broadcast address x.x.x.255 ? */
```

Comment:
The loopback packets might also get dropped here.

Reply:
As Tony pointed out this will drop the loopback packets as well. However, it also makes sense not to have the ( pxNetworkBuffer->pxEndPoint == NULL ) check at line 317. So, for me a good solution would be to change "xBadIPv4Loopback" at line 309 to "xIPv4Loopback" and then inside that have a check for "xBadIPv4Loopback", this way the loopback packets are not dropped as well.

else if( xIPv4Loopback( &( pxIPPacket->xIPHeader ) ) == pdTRUE )
{
    if( xBadIPv4Loopback( &( pxIPPacket->xIPHeader ) ) == pdTRUE )
    {
        /* The local loopback addresses must never appear outside a host. See RFC 1122
             * section 3.2.1.3. */
            eReturn = eReleaseBuffer;
    }
}

Reply:
@tony-josi-aws wrote:
> The loopback packets might also get dropped here.

That is a good observation

I agree with @shubnil. When I added the function `xBadIPv4Loopback()`, I didn't like the word "bad" because it is a negation. So yes I agree with his suggestion.
Let's call it `xAllowIPv4Loopback()`

Reply:
@tony-josi-aws Thanks for catching that. I completely forgot about loopback packets.
After considering you comment, I was going to substitute
```
( pxNetworkBuffer->pxEndPoint == NULL ) &&
```
with 
```
( xIsIPv4Loopback( ulDestinationIPAddress ) == pdFALSE ) &&
```
Let's call this "Option 1" It should work because the `if( xBadIPv4Loopback( &( pxIPPacket->xIPHeader ) ) == pdTRUE )` above check already took care of the situations where only the source OR the destination was a loopback address, so we can get away with just checking the destination.

@shubnil I like your idea because it keeps the checks a little better organized by keeping all loopback-related checks in one spot.
If I understand you correctly, you are suggesting that I create a new function:
```
BaseType_t xIPv4Loopback( const IPHeader_t * const pxIPHeader )
{
    if( xIsIPv4Loopback( pxIPHeader->ulDestinationIPAddress ) ||
        xIsIPv4Loopback( pxIPHeader->ulSourceIPAddress ) )
    {
        /* Either the source or the destination address is an IPv4 loopback address. */
        xReturn = pdTRUE;
    }

    return xReturn;
}
```
Let's call this "Option 2" It is better organized, but requires a new function and feels a little more wasteful because of the added function calls.

Instead of adding a function, I could just rewrite the else if() clause as such:
```
else if( xIsIPv4Loopback( &( pxIPPacket->xIPHeader.ulDestinationIPAddress ) ) == pdTRUE ||
               xIsIPv4Loopback( &( pxIPPacket->xIPHeader.ulSourceIPAddress ) ) == pdTRUE )
{
	/* source OR destination is a loopback address. Make sure the BOTH are */
	if( xBadIPv4Loopback( &( pxIPPacket->xIPHeader ) ) == pdTRUE )
	{
		/* The local loopback addresses must never appear outside a host. See RFC 1122
		* section 3.2.1.3. */
		eReturn = eReleaseBuffer;
	}
}
```
Let's call this "Option 3" It is a little harder to read, but there is no new function and all loopback tests are in one spot. Also, for most non-loopback packets the compiler can probably skip ahead after just one check.

Guys, I'm actually perfectly fine with all 3 options and I don't have much of a preference for any one of them. Let me know which one you want me to implement.

@htibosch 
If I rename `xBadIPv4Loopback()` to `xAllowIPv4Loopback()` I must also invert its returns. The long if()-elseif() clause in `prvAllowIPPacketIPv4()` gets messed up.
I would have to change the elseif() to check for pdFALSE, and also make `xAllowIPv4Loopback()` return pdTRUE for good loopback packets **as well as** all non-loopback packets and I don't like that last part.


p.s. I will do some actual test with loopback packets next week. Thank you ALL for the feedback!

Reply:
Thanks @evpopov  for sharing the various.
Option 3 looks the most clean and crisp solution to me.

Thanks,
Shub